base/system/services/database.c

```diff
@@ -621,9 +621,50 @@ ScmCreateOrReferenceServiceImage(PSERVICE pService)
 VOID
 ScmRemoveServiceImage(PSERVICE_IMAGE pServiceImage)
 {
+    DWORD dwError;
+
     DPRINT1("ScmRemoveServiceImage() called\n");
 
-    /* FIXME: Terminate the process */
+    /*
+     * No services are running in this image anymore.
+     * Tell the service dispatcher to exit now, and wait
+     * for the process to cleanly terminate.
+     */
+// FIXME: Re-enable here and remove the L"" SERVICE_CONTROL_STOP call in rpcserver.c
+#if 0
+    dwError = ScmControlService(pServiceImage->hControlPipe,
+                                L"",
+                                NULL,
+                                SERVICE_CONTROL_STOP); // Windows: use SERVICE_STOP
+    if (dwError == ERROR_SUCCESS)
+#endif
+    {
+        dwError = WaitForSingleObject(pServiceImage->hProcess, 20000);
+        if (dwError != WAIT_OBJECT_0)
+            DPRINT1("WaitForSingleObject failed, going to kill the process.\n");
+    }
+#if 0
+    else
+    {
+        DPRINT1("ScmControlService failed, going to kill the process.\n");
+    }
+#endif
+
+    if (dwError != ERROR_SUCCESS || dwError != WAIT_OBJECT_0)
```

Comment:
Do you mean
```suggestion
    if (dwError != ERROR_SUCCESS && dwError != WAIT_OBJECT_0)
```
?

Reply:
No. Check the two code paths above where the "going to kill the process" message is also emitted.

Reply:
Ok, your code _works_, as `ERROR_SUCCESS == 0 == WAIT_OBJECT_0`.
Nonetheless, it does _look_ wrong (as if always true), as checking twice for the same value (assuming reader knows the actual values are identical).
I suggest to add an explicit comment, or something.

Reply:
The fact they are equal is a coincidence, here.
The code that is above:
```c
    if (dwError == ERROR_SUCCESS)
    {
        dwError = WaitForSingleObject(pServiceImage->hProcess, 20000);
        if (dwError != WAIT_OBJECT_0)
            DPRINT1("WaitForSingleObject failed, going to kill the process.\n"); // Path n°2
        // else (dwError is == WAIT_OBJECT_0): Path n°3
    }
    else
    {
        DPRINT1("ScmControlService failed, going to kill the process.\n");
        // (dwError != ERROR_SUCCESS): Path n°1
    }

    if (the_test_in_question)
    {
        kill_the_process(pServiceImage->hProcess);
    }
```
There, I want to take the branch that kills the process, in case Path n°1 or Path n°2 were taken.

Reply:
> The fact they are equal is a coincidence, here.

Well, more than that: if values were different, then your `||` test result would always be `true`.

> what would be your suggestion then?

My suggestions are something like
```c
/* Usual '&&', implicit code redundancy */
if (dwError != ERROR_SUCCESS && dwError != WAIT_OBJECT_0)

/* No code redundancy, explicit comment */
if (dwError != 0) // 0 == ERROR_SUCCESS == WAIT_OBJECT_0.

/* Or, you could remove this 'if' and use a 'goto' on Path 3 to skip this kill block */
```